esphome/components/sonoff_d1/sonoff_d1.cpp

```diff
@@ -0,0 +1,304 @@
+
+  Permission is hereby granted, free of charge, to any person obtaining a copy of this software
+  and associated documentation files (the “Software”), to deal in the Software without
+  restriction, including without limitation the rights to use, copy, modify, merge, publish,
+  distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom
+  the Software is furnished to do so, subject to the following conditions:
+
+  The above copyright notice and this permission notice shall be included in all copies or
+  substantial portions of the Software.
+
+  THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
+  BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+  NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+  DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+  -----
+
+  If modifying this file, in addition to the license above, please ensure to include links back to the original code:
+  https://jeffresc.dev/blog/2020-10-10
+  https://github.com/JeffResc/Sonoff-D1-Dimmer
+  https://github.com/arendst/Tasmota/blob/2d4a6a29ebc7153dbe2717e3615574ac1c84ba1d/tasmota/xdrv_37_sonoff_d1.ino#L119-L131
+
+  -----
+*/
+
+/*********************************************************************************************\
+ * Sonoff D1 dimmer 433
+ * Mandatory/Optional
+ *  ^  0  1  2  3  4  5  6  7  8  9  A  B  C  D  E  F 10
+ *  M AA 55                                              - Header
+ *  M       01 04                                        - Version?
+ *  M             00 0A                                  - Following data length (10 bytes)
+ *  O                   01                               - Power state (00 = off, 01 = on, FF = ignore)
+ *  O                      64                            - Dimmer percentage (01 to 64 = 1 to 100%)
+ *  O                         FF FF FF FF FF FF FF FF    - Not used
+ *  M                                                 6C - CRC over bytes 2 to F (Addition)
+\*********************************************************************************************/
+#include <cmath>
+#include "sonoff_d1.h"
+
+namespace esphome {
+namespace sonoff_d1 {
+
+static const char *const TAG = "sonoff_d1";
+
+uint8_t SonoffD1Output::calc_checksum_(const uint8_t *cmd, const size_t len) {
+  uint8_t crc = 0;
+  for (int i = 2; i < len - 1; i++) {
+    crc += cmd[i];
+  }
+  return crc;
+}
+
+void SonoffD1Output::populate_checksum_(uint8_t *cmd, const size_t len) {
+  // Update the checksum
+  cmd[len - 1] = this->calc_checksum_(cmd, len);
+}
+
+char *SonoffD1Output::dump_cmd_(const uint8_t *cmd, const size_t len) {
+  static char dump[20 * 3 + 1] = {0};
+  auto ptr = &dump[0];
+  for (unsigned int i = 0; i < len; i++) {
```

Comment:
The type should match the max value you are comparing to.
```suggestion
  for (size_t i = 0; i < len; i++) {
```